mlfinlab/tests/test_labeling_tail_sets.py

```diff
@@ -0,0 +1,90 @@
+# pylint: disable=missing-module-docstring
+# pylint: disable=protected-access
+
+import unittest
+import os
+import numpy as np
+import pandas as pd
+from mlfinlab.labeling.tail_sets import TailSetLabels
+
+
+class TestTailSets(unittest.TestCase):
+    """
+    Unit tests for the tail sets labeling class.
+    """
+
+    def setUp(self):
+        """
+        Set the file path for the sample data.
+        """
+        project_path = os.path.dirname(__file__)
+        self.path = project_path + '/test_data/stock_prices.csv'
+        self.data = pd.read_csv(self.path, index_col='Date', parse_dates=True)
+
+    def test_init(self):
+        """
+        Tests to ensure exceptions are correctly raised for invalid inputs.
+        """
+        # Too large n_bins.
+        with self.assertRaises(Exception):
+            TailSetLabels(self.data[:100], n_bins=200)
+        # Window isn't int with vol_adj.
+        with self.assertRaises(Exception):
+            TailSetLabels(self.data[:100], n_bins=10, vol_adj='stdev', window='str')
+        # Window is too large
+            TailSetLabels(self.data[:100], n_bins=10, vol_adj='stdev', window=200)
+
+    def test_vol_adj_ret(self):
+        """
+        Tests for getting the correct volatility adjusted return.
+        """
+        # Check for no vol adjustment. Should be same as straight log returns.
+        label1 = TailSetLabels(self.data[:100], n_bins=10, vol_adj=None, window=None)  # No vol adjustment
+        test1 = label1.vol_adj_rets
+        simple_returns = np.log(self.data[:100]).diff().dropna()
+
+        # # Stdev adjusted returns.
+        label2 = TailSetLabels(self.data[:100], n_bins=8, vol_adj='stdev', window=20)
+        test2 = label2.vol_adj_rets
+        test2_actual = (simple_returns / simple_returns.rolling(20).std()).dropna()
+
+        # Mean abs returns
+        label3 = TailSetLabels(self.data[:100], n_bins=12, vol_adj='mean_abs_dev', window=20)
+        test3 = label3.vol_adj_rets
+        pd.testing.assert_frame_equal(test1, simple_returns)
+        pd.testing.assert_frame_equal(test2, test2_actual)
+        np.testing.assert_array_almost_equal(test2.iloc[2:3, 0:10],
+                                             np.array([[0.13093775, -0.04441354, -1.18762953, -0.18278121, -0.61415685,
+                                                        -1.02704163, -0.47273006, -0.34072258, -0.73818837,
+                                                        0.03212328]]))
```

Comment:
Why are we testing `test2` values again if we have compared them to `test2_actual` already?